packages/core/src/browser/widgets/select-component.tsx

```diff
@@ -84,8 +80,22 @@ export class SelectComponent extends React.Component<SelectComponentProps, Selec
         this.dropdownElement = list;
     }
 
+    protected resolveSelected(props: SelectComponentProps): number {
+        let selected = 0;
+        if (typeof props.value === 'number') {
+            selected = props.value;
+        } else if (typeof props.value === 'string') {
+            selected = Math.max(props.options.findIndex(e => e.value === props.value), 0);
+        }
+
+        return selected;
+    }
+
+    override componentWillReceiveProps(nextProps: SelectComponentProps): void {
```

Comment:
This class is a bit of an odd duck: it's a React component, but it has also been designed to work in non-React contexts (`PreferenceSelectInput`). Because of that, it's been designed to behave more like an unmanaged input rather than a managed input, to behave more like a vanilla DOM input rather than a typical React input. And I believe that is correct - for it to be useable outside React contexts, it _must_ behave more like an unmanaged input, and be built for use with `Refs`. Given that, its state should in general not be managed by changing `props` but imperatively; the `value` field should be treated as (and perhaps renamed to) `defaultValue` to indicate that it will take effect on initialization, but be ignored thereafter, and `selection` should be moved back into the `state` object: the actual selection should be managed by the component itself, and only updated externally imperatively. I believe moving `selection` from state to a class member is currently only working because every time it is set, it happens to cooccur with a different, React-y means of changing state: at construction time, in `willReceiveProps`, or coincidentally when the value of `state.hover` is also changed.

Reply:
I have have updated the handling of `state` as per your proposal, 
I just wanted to clarify that the `willReceiveProps` does not change the state,
it simply anticipates the reception of new props passed on from the parent component which then get to be used in the following call i.e. `render`